### PR TITLE
Fix "Light Only" mode of `CanvasItemMaterial`

### LINKED ISSUE #44559: Light-only mode in CanvasItemMaterial makes both sprite and light disappear
**Godot version:**

bccbd4b

**Issue description:**

2TU6IRUzXA

**Steps to reproduce:**
1. Create Sprite2D
2. Create Light2D
3. Assign icon.png to both
4. Add CanvasItemMaterial to Sprite2D node
5. Set the light mode in the material to light only

### PR DESCRIPTION
**Pull request by dalexeev**
Closes #44559.

### CODE REVIEW

File: servers/rendering/renderer_rd/shaders/canvas.glsl [hunk 1]
```diff
@@ -611,6 +611,9 @@ void main() {
 		}
 
 		light_blend_compute(light_base, light_color, color.rgb);
+#ifdef MODE_LIGHT_ONLY
+		ligth_only_alpha += light_color.a;
```
Comment (dalexeev) [inline servers/rendering/renderer_rd/shaders/canvas.glsl:615]: I'm not sure what should be used here, sum, average or something else. And is `clamp` needed here?

**Comment by dalexeev** [comment]
By the way, I noticed two more bugs:

1\. Blending works differently in 3.x and 4.0 (these are 2 different projects, but it is already clear that it works differently):



2\. In 4.0, the alpha channel of the `Light2D.color` property is ignored.

**Comment by chrom007** [comment]
Finally someone solved this problem! Thank you!

**Comment by dalexeev** [comment]
Test projects for comparison with 3.x:

[light-only-3.x.zip](https://github.com/godotengine/godot/files/11128831/light-only-3.x.zip)
[light-only-4.0.zip](https://github.com/godotengine/godot/files/11128830/light-only-4.0.zip)

Note that background and canvas modulate affect the result. The difficulty is that the Light Only mode must take into account the background color.

**Comment by clayjohn** [comment]
This is on the right track, but it is still drawing the original sprite. The "Light Only" mode in 3.x only drew the contribution from the Light2D

_3.x: notice how the original sprite looks very faint?_
Screenshot from 2023-04-11 14-16-05

_This PR: notice how the Light2D is essentially only acting as a mask_
Screenshot from 2023-04-11 14-16-01

This PR is a step in the right direction. But the trouble is that the final color is ultimately mixed with the background using the blend mode of the Sprite rather than the blend mode of the light (which is kind of necessary with how "Light Only" worked. 

It would be interesting to hear from people like @KoBeWi to see if this PR would be enough for their use case as I think matching the 3.x behaviour exactly may be too difficult

**Comment by KoBeWi** [comment]
Well, my use-case is using lights to reveal hidden things (objects, outlines). The lights and hidden objects are using exclusive `light_mask` and the light texture is flat. Works perfectly with this PR.

https://user-images.githubusercontent.com/2223172/232028474-5038123e-f2fd-4ab8-8f9f-9b6d79b9085b.mp4

**Comment by dalexeev** [comment]
I marked this PR as draft for two reasons:

1. The issue is not properly fixed, although as clayjohn pointed out, even this rough fix might be enough for basic scenarios.
2. I planned to investigate the problem, but it's been almost a month and during this time I haven't touched it. I don't think I'll come up with any new ideas unless someone more experienced in rendering gives me a hint.

So I undraft this PR. If someone has the desire and confidence to fix it properly, then I don't mind. Let us know about it and feel free to use it as a basis.

Or we can merge this PR as is, for the reason "it's better to have a feature that doesn't work well with advanced scenarios than a feature that doesn't work at all" and fix it properly in future PRs.

**Comment by Janonas** [comment]
Would you be willing to share your compiled exe file with these fixes? Light only mode is pretty integral for my project, and i cannot continue development without it working properly. Ive tried compiling the code myself, but i must be doing something horribly wrong, as i cannot seem to get it working. It would be a huge help as the only other alternative i have is manualy porting the project back to godot 3, which would be a major headache.

**Comment by dalexeev** [comment]
> Would you be willing to share your compiled exe file with these fixes?

I rebased and force-pushed the branch. When the build is over you will be able to download the artifacts.

**Comment by Janonas** [comment]
Thank you, it works flawlessly for my purposes.

**Review by clayjohn — APPROVED** [review]
Let's go ahead with this. The users who are missing a "light only" mode seem to all be happy with this approach. I'm not sure that fully replicating the 3.x behaviour is possible/desirable and this approach makes a lot of sense given how 2D light rendering works in Godot 4

**Comment by YuriSizov** [comment]
Sounds good. Thanks!

**Pull request merged**